[4-6단계 - 방탈출 사용자 예약] 멍구(이유영) 미션 제출합니다.

YuyoungRhee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [x] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

궁금한 점은 코멘트로 남겼습니다!

추가적으로 패키지 구조를 변경했습니다. 도메인 first 구조로 패키지를 분리하였는데, 개인적으로 계층first 구조에서 코드를 찾기가 힘들어 변경해보았습니다.

### 1-3단계 코멘트 답변
**1. gradle 버전 설정을 바꾼 이유**
Java 21버전은 Gradle 8.5부터 지원한다고 합니다. 어떤 이유인지 당시에 테스트 하나만 Java 21 버전 Compile 오류로 동작하지 않았습니다. 따라서 Gradle 버전을 맞추어 제대로 작동하도록 변경했습니다.
- 인텔리제이 설정, gradle이 동작하는 지 정확하게 잘 모르겠으나 우선 이렇게 해결하고 넘어갔습니다..! gradle에 대해서는 시간이 남는다면 공부할 수 있을 것 같습니다.

**2. ReservationPolicy로 조건을 분리한 것에 대해**
> 다만, Reservation이 생성될 때 꼭 필요한 조건들이라면 도메인 내부에서 해주는게 더 적합할 수도 있겠다는 생각이 들기도 해요.
> 누군가 Reservation을 만들 때, 해당 Policy를 적용시키는 코드를 빼먹는다면 정책이 반영되지 않은 예약이 비즈니스 코드에서 사용될 수 있는게 우려점으로 남는 것 같아요.
> 이 부분에 대해서 어떻게 생각하시나요?~

인정합니다! 다만 저는 모든 Reservation에 대해서 항상 필요한 조건이라고 생각하지 않습니다. 예를 들어 관리자는 조금 늦은 손님에게 직접 예약을 해주고 운영해볼 수 있을 것 같아요. 관리자가 1분 늦은 예약을 생성할 때 예외가 터진다면 곤란할 것 같습니다. 너무 자세한 예시인 것 같지만, 이런식으로 Reservation 생성 행위 자체에 대한 규약은 언제든지 바뀔 수 있다고 보기 때문에  "Reservation이 생성될 때 꼭 필요한 조건" 이라고 보지는 않고 있어요.

**3. mock을 활용한 테스트**
> mock을 활용하게 되면 사실상 의도한 대로 동작하는지 테스트하는거라 테스트코드가 의미없게 느껴지기도 하는 것 같아요. 좀 더 확실히 하기 위해서는 verify 등을 활용해서 실제로 호출했는지를 검증해보는 것은 어떨까요?

이번 테스트를 작성하면서 의식해보았는데, 의미 없는 테스트가 작성됐다면 코멘트 부탁드려요!!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

YuyoungRhee committed:
feat(static): 사용자에게 오류 원인을 보여줄 수 있도록 추가

YuyoungRhee committed:
refactor(test): service 테스트 verify를 이용한 호출 검증 추가

YuyoungRhee committed:
fix(Fixture): 테스트 Fixture에서 id가 동일한 문제 해결

YuyoungRhee committed:
refactor: 정적 view 전용 Controller 제거 후 viewController 등록으로 대체

YuyoungRhee committed:
refactor: 패키지 구조 변경, reservation 하위 패키지 생성

YuyoungRhee committed:
fix: 인기테마 페이지에서 썸네일이 출력되지 않는 문제 해결

- 썸네일 이미지 추가, data.sql에서 해당 썸네일을 참조하도록 함
- style.css에서 썸네일 이미지의 최대 세로 크기 제한

YuyoungRhee committed:
feat(auth): 사용자 로그인 api 구현

- AuthService, JwtTokenProvider 등 구현
- 에러 처리는 AuthorizationException 커스텀예외를 통해 처리
- users 관련 테이블 스키마 및 데이터 추가

YuyoungRhee committed:
refactor(exception): 예외 관련 패키지 분리

- 루트 exception 패키지와 reservation/exception 패키지로 분리

YuyoungRhee committed:
feat(auth): 사용자 인증 정보 조회 api 구현

YuyoungRhee committed:
fix(auth): 프론트 4단계 주석 해제 - 사용자 정보 체크

YuyoungRhee committed:
fix(TokenLoginController): cookie 설정 추가 세팅

- HttpOnly, Path 설정 추가: xss 위협 방지, 쿠키 경로 설정 추가

YuyoungRhee committed:
refactor(auth): JJWT deprecated API 제거 및 최신 API로 변경

YuyoungRhee committed:
refactor(auth): HandlerMethodArgumentResolver로 사용자 인증 로직 분리

- Cookie에서 토큰 추출 후 사용자 객체로 매핑하는 책임을 ArgumentResolver로 이전
- 컨트롤러에서는 @AuthenticatedUser로 도메인 User를 직접 주입받도록 변경

YuyoungRhee committed:
fix(auth): 토큰이 없는 경우 발생하는 예외 처리

- Cookie가 null인 경우 로그인 토큰 자체가 없기 때문에 로그인이 필요한 상황 -> 로그인 필요 예외로 처리

YuyoungRhee committed:
refactor(reservation): 사용자가 예약 생성 시, 로그인한 사용자 정보를 활용하도록 리팩터링

- Argument Reolver를 활용하여 token에서 사용자 정보를 추출해 주입
- 프론트 코드 변경

YuyoungRhee committed:
refactor: user -> member로 네이밍 전체적 변경

YuyoungRhee committed:
fix: 관리자페이지에서 예약 등록이 가능하도록 수정

YuyoungRhee committed:
refactor: 패키지 구조 도메인별로 분리

YuyoungRhee committed:
feat: 어드민 페이지 진입은 admin 권한만 가능하도록 구현

YuyoungRhee committed:
feat: 예약 목록 검색 기능 구현

YuyoungRhee committed:
refactor: dto 패키지 구조 변경

YuyoungRhee committed:
fix: 도메인 객체에서 id가 null인 상태에서 동일성을 보장하지 않도록 변경

YuyoungRhee committed:
fix(ReservationSearchCondition): memberName으로 검색이 아닌 memberId 검색으로 변경

YuyoungRhee committed:
refactor(JdbcReservationQueryDao): 일관성에 맞게 findById 시 Optional을 반환하도록 수정

YuyoungRhee committed:
refactor(AuthErrorCode): 에러 문구 수정

YuyoungRhee committed:
refactor(ReservationDetailResponse): 중첩 레코드 접근 제한자를 private으로 변경

YuyoungRhee committed:
test: 전체적인 테스트 추가 및 개선

YuyoungRhee committed:
refactor(JdbcReservationQueryDao): NamedParameterJdbcTemplate + SqlBuilder 도입으로 동적 쿼리 가독성 개선

YuyoungRhee committed:
docs: REST Docs를 이용한 API 명세 작성

YuyoungRhee committed:
refactor(exception): CustomException 예외 공통 처리

YuyoungRhee committed:
refactor(auth): Argument Resolver에서 Member 대신 LoginMember Dto를 주입하도록 변경

- 불필요한 정보를 넘기지 않도록 함 (password, role)

YuyoungRhee committed:
docs: 리드미 업데이트, 테스트 계정 안내 추가

yuyoungrhee commented on src/main/java/roomescape/auth/presentation/AdminAuthorizationInterceptor.java:
```diff
@@ -0,0 +1,54 @@
+        this.jwtTokenProvider = jwtTokenProvider;
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+            throws IOException {
+
+        Optional<String> result = extractor.extract(request);
+
+        if (result.isEmpty()) {
+            return handleRedirectWithAlert(response, "로그인이 필요합니다.", "/login");
+        }
+
+        String token = result.get();
+        if (!jwtTokenProvider.validateToken(token)) {
+            return handleRedirectWithAlert(response, "유효하지 않은 토큰입니다.", "/login");
+        }
```
RESTful API 환경에서는 인증 예외 발생 시 JSON으로 응답을 내려주고, 클라이언트가 이를 해석해 처리하는 방식이 일반적이라 생각합니다.
따라서 `AuthenticationException`을 발생시키고 프론트 쪽에서 처리해보려고 했으나 js 코드 작성 능력이 부족하여 우선 서버에서 직접 처리하도록 구현했습니다.

seovalue replied:
오홍 자체적으로 리다이렉팅 되게 해주신거군요?

yuyoungrhee replied:
네 맞습니다! 그런데 Interceptor에서는 예외만 발생시키고, ExceptionResolver에서 예외를 처리하는 방식이 좀 더 책임이 의미있게 분리되는 거 같아 개선해봤습니다.

[refactor(auth): Interceptor 예외 직접 처리 방식 → Resolver 위임 방식으로 개선](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/da448f87be4af7e37821635cae2f548244b1e37a)

yuyoungrhee commented on src/main/java/roomescape/config/WebMvcConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+    private final AuthenticatedMemberArgumentResolver resolver;
+    private final AdminAuthorizationInterceptor adminInterceptor;
+
+    public WebMvcConfiguration(AuthenticatedMemberArgumentResolver resolver, AdminAuthorizationInterceptor adminInterceptor) {
+        this.resolver = resolver;
+        this.adminInterceptor = adminInterceptor;
+    }
+
+    @Override
+    public void addViewControllers(ViewControllerRegistry registry) {
+        // 사용자 페이지
+        registry.addViewController("/").setViewName("index");
+        registry.addViewController("/reservation").setViewName("reservation");
+
+        // 관리자 페이지
+        registry.addViewController("/admin").setViewName("admin/index");
+        registry.addViewController("/admin/reservation").setViewName("admin/reservation-new");
+        registry.addViewController("/admin/time").setViewName("admin/time");
+        registry.addViewController("/admin/theme").setViewName("admin/theme");
+    }
```
정적 뷰 페이지는 별도의 로직이 없기 때문에, 직접 컨트롤러 클래스를 작성하는 대신 Spring의 `addViewControllers()`를 활용해 뷰 경로를 등록했습니다. 정적 페이지와 관련된 매핑을 한 곳에서 관리할 수 있고, 불필요한 코드 작성을 줄여준다는 장점이 있는 것 같습니다.

seovalue replied:
오호! 원래 알고 있던 Registry셨나요?
해당 Registry에 대해서 좀 더 설명해주실 수 있으실까요?

yuyoungrhee replied:
이번 미션을 진행하기 위한 MVC Config 학습 테스트에서 새로 알게 됐습니다.

ViewControllerRegistry는 “URL 요청 경로 → View 이름”을 매핑해주는 정보를 등록하는 객체입니다.

**ViewControllerRegistry를 통한 자동 등록**
1. 개발자가 `addViewController()`로 URL 경로와 뷰 이름을 등록
2. 해당 매핑 정보가 `ViewControllerRegistration` 객체에 저장됨
3. 실행 시 ViewControllerRegistration 정보를 바탕으로, URL → ViewController 매핑이 `RequestMappingHandlerMapping`에 등록됨
4. 클라이언트 요청 시, `ViewController`가 `ModelAndView("뷰 이름")`을 반환
5. `ViewResolver`가 뷰 이름을 해석하여 최종 View를 렌더링함

**`@Controller`를 통한 수동 등록**
1. Spring이 `@Controller` 클래스를 Component 스캔하여 Bean으로 등록
2. 해당 클래스 내 `@RequestMapping`,` @GetMapping` 등 애너테이션이 붙은 메서드를 분석
3. URL → 메서드 매핑 정보를 `RequestMappingHandlerMapping`에 등록
4. 클라이언트 요청 시, 해당 메서드가 호출되어 뷰 이름을 반환
5. `ViewResolver`가 뷰 이름을 해석하여 최종 View를 렌더링함

seovalue replied:
좋네요! 새롭게 알게된 건 크루들이랑도 나누면서 체화하면 좋을 것 같아요!

yuyoungrhee commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -10,51 +10,31 @@
         detail.setDetail(code.getMessage());
         detail.setProperty("code", code.name());
         return ResponseEntity.status(code.getStatus()).body(detail);
     }
```
기존: 비즈니스 예외 / NotFound 예외 / 입력 예외 등 세부 예외 별 `@ExceptionHandler` 작성
변경: 가장 상위 부모 추상클래스인 `CustomException`만 핸들링하도록 통합

- 불필요한 코드 중복을 줄이고 GlobalExceptionHandler에서 너무 세부적인 것을 신경쓰지 않기 위해 리팩터링했습니다. 
- 예외 추가/변경 시 GlobalExceptionHandler의 수정 없이 처리 할 수 있는 구조로 개선했습니다.

yuyoungrhee commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -1,35 +1,37 @@
 public class Reservation {
     private final Long id;
-    private final String name;
+    private final Long memberId;
```
Member에 대한 정보는 직접참조가 아닌 ID를 통한 참조로 구현했습니다.

시간과 테마는 `방탈출 예약`을 온전히 구성하기 위한 정보이기 때문에 객체를 소유하도록 자연스럽게 설계했습니다. 그러나 유저는 예약의 소유자이지, 의미적으로 내부에 반드시 포함해야 하는 구성요소로 보이지는 않았습니다. 따라서 id 참조가 조금 더 자연스럽다고 생각해 ID참조로 구현했습니다.

이렇게 설계하면서  직접 참조와 식별자 참조의 경계를 명확히 하기 위한 기준이 필요함을 느꼈습니다. 지금까지는 이렇게 "느낌"적으로만 구현했는데, 정확한 기준을 세우기 위해서는 어떤 키워드를 학습해보면 좋을까요?

seovalue replied:
예를 들어 Reservation 조회 시 Member의 email, name, role 등과 같이 member 도메인의 정보가 항상 응답되어야한다면 id 보다는 직접 객체를 갖고 있는 형태가 좀 더 용이할 것 같아요. 그 외의 케이스는 간접참조로 둬도 좋을 것 같고요!ㅎㅎ
우선 지금 단계에서는 이정도만 학습해둬도 좋을 것 같네요~

yuyoungrhee replied:
감사해요! 지금은 member 도메인 정보가 항상 같이 응답되어야하는데, ID 참조로 둔 형태가 됐군요. 이를 개선하려면 크게 구조를 바꿔야할 것 같아서, 다음 요청까지는 유지하겠습니다. 다른 코드들이 정리되고 나면 다시 직접참조로 바꿀지 결정해보도록 할게요!

yuyoungrhee commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -70,14 +72,20 @@ public boolean equals(Object o) {
-        if (!(o instanceof Reservation that)) {
+        if (!(o instanceof Reservation other)) {
             return false;
         }
-        return Objects.equals(id, that.id);
+        if (this.id == null || other.id == null){
+            return false;
+        }
+        return Objects.equals(id, other.id);
```
엔티티의 ID가 아직 부여되지 않은 경우, equals 비교 자체가 잘못된 시도라고 판단해 false를 반환하도록 했습니다.

다만 한 가지 고민은, 이런 상황이 사실상 로직 오류에 가깝다면 equals에서 아예 예외를 던지는 것이 더 명확한 신호일 수 있지 않을까 하는 점입니다.

equals를 호출했을 때 예외를 기대하지 않기 때문에 외부 호출자 입장에서 항상 기대되는 값(boolean)을 안정적이게 얻기 위해서 false를 던지는 것이 조금 더 낫다고 생각합니다. 그러나 fast-fail 관점에서는 에외가 낫지 않을까 싶습니다.

장단점이 있는 것 같은데, 조앤의 생각을 공유해주실 수 있나요?

seovalue replied:
일반적인 라이브러리 호출 안정성을 우선한다면, equals() 내부에서 예외를 던지기보다 ID 미할당 시 무조건 false 반환이 더 무난한 것 같아요. 
다만 “ID 없는 엔티티 동등성 비교가 곧 로직 오류”라는 점을 별도의 검증 레이어나 테스트에서 Fail-Fast로 잡아내도록 설계할 수 있을 것 같아요. 이런 경우에 대한 비교가 불가능하도록 미리 잡아내는거죠.
혹은 엔티티 고유성을 보장해 줄 비즈니스 키를 equals() 기준으로 채택하는 방법도 있을 것 같긴 한데, 지금은 entity가 거의 schema와 같은 역할이여서 이건 어색할 수도 있을 것 같기도 하고요 🤔

yuyoungrhee replied:
충분히 이해됐습니다..! 만약 ID가 없는 상태에서의 동등성 비교가 로직 오류라면, equals에서 예외를 던지기보다는 해당 상황이 발생하지 않도록 검증 레이어나 테스트에서 사전에 잡아내는 쪽이 더 적절해 보이네요😀

yuyoungrhee commented on src/main/java/roomescape/reservation/infrastructure/JdbcReservationRepository.java:
```diff
@@ -0,0 +1,54 @@
+        jdbcTemplate.update(deleteSql, id);
+    }
+
+    @Override
+    public boolean existsDuplicatedReservation(LocalDate date, Long timeId, Long themeId) {
+        String sql = """
+            SELECT EXISTS (
+                SELECT 1
+                FROM reservation
+                WHERE theme_id = ? AND time_id = ? AND date = ?
+            )
+        """;
+        return jdbcTemplate.queryForObject(sql, Boolean.class, themeId, timeId, date);
```
<날짜, 시간, 테마가 중복되는지 여부 검사 로직>
기존: 모든 List<Reservation>을 불러와 Service에서 자바 코드 로직으로 처리
변경: SQL문으로 처리

자바코드로 처리했을 때는 비즈니스로직을 온전히 코드로 풀어낸다는 점에서 명확성, 가독성이 좋았지만 성능문제가 우려됐습니다. 특히 filter로 하나씩 검사하는 구조이기 때문에  쉽게 성능 이슈를 예상할 수 있었습니다. 이또한 트레이드오프 영역 같습니다만, **메서드명과 파라미터 네이밍으로 어느정도 예상 가능한 코드인 것 같아 SQL문으로 로직을 풀어냄을 선택**했습니다.

yuyoungrhee commented on build.gradle:
```diff
@@ -28,8 +39,42 @@ dependencies {
+    runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.11.5'
+    runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.11.5'
+
+    asciidoctorExt 'org.springframework.restdocs:spring-restdocs-asciidoctor'
```
step1 때 API 명세를 작성해보라는 제안을 해주셨는데, 작성하기 시작하니 막막하더라구요,,ㅎㅎ
Swagger와 Rest docs 등의 존재를 알고 있었는데, 이미 RestAssured로 모든 API에 대한 테스트 작성이 끝나서 RestDocs를 쉽게 활용할 수 있을 것 같았어요. 실제로 given 절 정도만 변경한 뒤 문서화할 수 있었습니다!

seovalue replied:
와우 멋집니다!

yuyoungrhee commented on src/main/java/roomescape/reservation/exception/ImpossibleReservationException.java:
```diff
@@ -0,0 +1,9 @@
+package roomescape.reservation.exception;
+
+import roomescape.exception.DomainException;
+
+public class ImpossibleReservationException extends DomainException {
+    public ImpossibleReservationException(String message) {
+        super(message, "예약 불가", "IMPOSSIBLE_RESERVATION");
+    }
+}
```
현재 DomainException이라는 공통 상위 예외 클래스를 root/exception 패키지에 두고,
각 도메인별 예외(ImpossibleReservationException 등)에서 이를 상속받는 구조로 설계했습니다.

그런데 이로 인해 도메인(reservation)이 글로벌 성격의 exception 패키지를 직접 의존하게 되어,
도메인이 외부를 참조하게 되는 구조가 만들어졌습니다.

저는 도메인은 가능한 외부에 무지해야 하고, 캡슐화되어야 한다고 생각하는데요.
이 구조가 계층 분리나 독립성 측면에서 문제가 될 수 있을까요?

혹시 이를 해결할 수 있는 구조나 관점이 있다면 조언 주시면 감사하겠습니다.
1. 도메인에서 root/exception을 바라보는 건 괜찮은가?
2. 만약 괜찮지 않다면, 어떻게 해야 이 의존을 끊을 수 있을까?
3. 의존을 끊을 수 없다면 설계 자체를 다시 재고해보아야할까?

seovalue replied:
root에 위치하는 Exception들과 도메인 내부에 위치하도록 해주신 Exception은 어떤 차이예요?
그리고 멍구는 어떤 방식을 고민해보셨어요?

yuyoungrhee replied:
### `root/exception/` 안의 Exception과 `root/reservation/exception/` 안의 Exception 차이
root 패키지의 Exception은 표현영역, 응용영역에서 의존하고 있습니다.

도메인 내부의 exception은 외부와 상관 없이 오로지 도메인 규칙에 의거한 예외를 던집니다. 이 예외가 어디서 어떻게 처리되는지는 관심 없습니다. 마치 순수 자바로 구현한 레벨1의 도메인 로직처럼 메시지를 던지고 외부로 전파만 합니다.

이때 Service에서 잡는 방법이 있고, ExceptionHandler로 잡는 방법이 있었습니다. 예외 처리라는 관심사를 GlobalExceptionHandler에 모으고자 도메인 예외 또한 GlobalExceptionHandler에서 잡는 방식을 선택했습니다.

### DomainException을 root에 둔 이유
Domain 관련 예외를 잡는 최상위 계층이 있었으면 했습니다. 아니라면 GlobalException에서 ReservationException, TimeException, ThemeExeption 등과 같이 도메인이 늘어날 때마다 새로 `@ExceptionHandler`를 통해 잡아주어야 합니다.

따라서 `DomainException`이라는 상위 추상예외클래스를 상속받아 각 도메인에서 구현하는 방식으로 설계하고 싶었습니다. 자바 기본 예외 클래스를 사용하는 방법도 있었지만, 이제 "이건 도메인 내부에서 발생하는 예외이다"를 네이밍으로 표현해주는 게 더 중요하다고 생각했어요.

### 개선 사항
정리하다보니 현재 고민하고 있는 도메인->외부 의존성을 해결하는 방법이 있는 거 같아 적용해봤습니다.
`DomainException`이 root에 있는 게 아니라 도메인에서 공통으로 사용하는 어떤 패키지를 만들어야할 것 같네요.. 현재는 root를 "모든 영역에서 사용 가능함"을 의미하고 있으니, Domain은 내부만 알길 원한다면 공통 도메인 로직을 모아두는 패키지로 나타내주면 좋을 것 같아요.

우선은 `shared/domain/exception` 루트로 이동해보았습니다. 여러가지로 찾아보고, 가장 일반적인 네이밍을 사용했어요!

[refactor: DomainException의 위치를 shared로 이동 ](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/d0b5c7f109cd7122f223953b260ce65280d3649b)

seovalue replied:
고민한 내용 전달주셔서 감사해요! 처음에 제 의견을 안드렸던 이유는 저는 우테코안에서 시행착오를 많이 겪은 뒤 본인만의 무언가를 만들어가는게 중요한 것 같다고 생각해서요ㅎㅎ 최대한 멍구가 생각해보시고, 그대로 구현했을 때 어떤 장/단점이 있는지 직접 겪어보시면서 본인만의 의견을 만들어가시길 기대해요!

yuyoungrhee replied:
좋은 말씀 감사합니다😲 시간에 쫓길 필요는 없지만 미션 마감 기한이 정해져있다보니 어쩔 수 없이, 시간이 많이 들겠다 싶은 변경 사항은 머릿속으로만 생각해보고 직접 구현해보지 않는 경우가 많은 것 같아요. 시간적 이익을 따져들지 말고 우선 직접 경험해보는 방법이 제일 좋은 것 같아요!! 앞으로 열심히 이것저것 삽질해보려고 합니다ㅎㅎ

YuyoungRhee committed:
feat: request dto에 대한 유효성 @Valid로 체크

m-a-king commented:
restdocs 너무 잘 봤습니다. 💯

yuyoungrhee commented on src/main/java/roomescape/reservation/application/ReservationQueryService.java:
```diff
@@ -0,0 +1,30 @@
+import roomescape.reservation.infrastructure.dto.ReservationDetailData;
+
+@Service
+public class ReservationQueryService {
```
### QueryService를 따로 만든 이유

Reservation 추가API에서 응답을 온전하게 반환하기 위하여 `memberService.getMemberById(memberId)`를 다시 호출해야하는 상황이 있었습니다.

예약 추가, 삭제 API의 경우 memberId만 관심사이지, member 자체가 아니었습니다. 즉 member를 가져오는 것은 service의 관심사가 아니라고 생각했습니다.

따라서 Reservation 안의 모든 정보를 반환하는 것은 "조회"의 관심사를 둔 다른 Service로 나누는 것으로 구현했습니다.
ID 참조가 아닌 직접참조라면 자연스럽게 모든 정보를 Reservation이 가지고있기 때문에 나누지 않았을 것 같은데, ID 참조를 하게 되면서 이와 같은 상황을 겪게된 것 같습니다.

**정리**
- 등록/삭제는 member의 ID만 알면 됨 → 도메인 로직의 관심사
- 응답에 필요한 member 정보 조합 → 표현 계층의 관심사
- 이 둘을 나누어서 책임을 분리함

seovalue replied:
그럼 발생하는 모든 read 쿼리는 queryService에서 다루게 되나요?
만약 findOrCreate와 같은 로직이 들어간다면 어디에 구현되어야할까요?

yuyoungrhee replied:
findOrCreate가 Controller 관점에서 원하는 기능이라고 한다면, service와 queryService 를 조합해서 구현할 수 있을 것 같습니다.

```java
//예시
Dto dto = service.createIfNotExists(인자);
queryService.getById(dto.id());
```

Member는 Reservation의 로직에서 다루는 게 아니라 조회할 때만 필요한 정보입니다.
- Reservation의 비즈니스적 로직이 들어갈 때는 항상 ReservationService를 사용
- 클라이언트에서 Member 정보를 포함한 응답을 원한다면 controller에서 queryService를 마지막에 사용

yuyoungrhee replied:
Member를 직접참조로 갖고 있었다면 QueryService는 필요하지 않겠네요. 이것 말고도 더 고민해봐야할 포인트들이 있는 거 같아서, [ID참조 vs 직접참조 고민 코멘트](https://github.com/woowacourse/spring-roomescape-member/pull/253#discussion_r2082389051) 에 남긴 답변처럼 다음 요청까지는 이 구조를 유지하겠습니다.

seovalue replied:
좋네요! 저도 QueryService를 한번 만들어본 적이 있는데 그 이후 findOrCreate가 들어오면 어디에 둬야할지 고민이 된 경험이 있었거든요. 어디에 둬야할지 고민이 되는 순간 허들이 된다고 생각해본 적이 있어서 의견을 여쭈었었어요! 👍

YuyoungRhee reviewed (commented):
리뷰 감사합니다 조앤! 코드 의도 설명과 궁금한 점을 코멘트로 남깁니다.

YuyoungRhee committed:
refactor(ReservationService): memberId의 존재 여부만 확인하도록 변경

YuyoungRhee committed:
build(gradle): 중복 설정 제거

seovalue commented on README.md:
```diff
@@ -1,6 +1,34 @@
+
+## API 명세서
+main 실행 후 -->
+[API 명세 보기](http://localhost:8080/docs/index.html)
```
👍 센스굿

seovalue replied:
👍 센스굿

seovalue commented on src/main/java/roomescape/auth/application/AuthService.java:
```diff
@@ -0,0 +1,58 @@
+import roomescape.member.domain.repository.MemberRepository;
+
+@Service
+public class AuthService {
```
AuthService의 역할은 뭐예요?

yuyoungrhee replied:
인증을 담당하는 서비스입니다! 현재는 토큰 생성과 인증 관련 역할을 하고 있습니다.

1. 토큰 생성
로그인 요청이 들어오면, `createToken()`에서 이메일 + 비밀번호를 검증한 후 JWT를 발급합니다.
2. JWT 파싱 및 검증
`findMemberByToken()`에서 토큰의 유효성 검사 후 사용자 정보를 조회합니다.

seovalue commented on src/main/java/roomescape/auth/dto/LoginMember.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.auth.dto;
+
+public record LoginMember(
```
👍 LoginMember라는 dto를 따로 만들어주시게 된 이유도 궁금해요~!

yuyoungrhee replied:
처음에는 Member 도메인을 컨트롤러에 직접 주입했습니다.  
하지만 이 방식은 컨트롤러가 password, role 등 민감한 정보에까지 접근할 수 있게 되어, 불필요하게 도메인 내부 구현에 의존하게 되는 문제가 있습니다.

Member는 시스템에서 가장 중요하게 보호되어야 할 도메인이므로, Member 유효성 검사는 Resolver에서 마치고, 컨트롤러에는 필요한 정보만 전달하는 것이 더 적절하다고 생각했습니다.

따라서, 컨트롤러에는 Member 전체 대신 필요한 필드만 포함한 Dto를 전달하도록 설계했습니다.
[Member에서 Dto로 변경한 커밋](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/34f5832ce378fe946e25a5abc70b4a63496f6a96)

seovalue commented on src/main/java/roomescape/auth/exception/AuthExceptionResolver.java:
```diff
@@ -0,0 +1,42 @@
+import roomescape.exception.ErrorCode;
+
+@Component
+public class AuthExceptionResolver implements HandlerExceptionResolver {
```
HandlerExceptionResolver에 대해서 설명해주세요! 사용하게 된 이유도요ㅎㅎ

yuyoungrhee replied:
**`HandlerExceptionResolver`를 사용하게 된 이유는 Interceptor에서 발생한 예외를 처리하기 위함입니다.**

Spring의 예외 처리 플로우에서 Interceptor는 컨트롤러 실행 이전에 동작하므로, `@ControllerAdvice`나 `@ExceptionHandler`로는 해당 예외를 처리할 수 없습니다. 대신 DispatcherServlet은 내부에서 예외가 발생하면 `processHandlerException()`을 호출하고,
등록된 `HandlerExceptionResolver` 리스트를 순차적으로 탐색하여 예외 처리를 시도합니다.

이 Resolver 리스트에는 다음과 같은 기본 처리기가 포함되어 있습니다:
- ExceptionHandlerExceptionResolver: `@ControllerAdvice`, `@ExceptionHandler` 처리
- ResponseStatusExceptionResolver: `@ResponseStatus`, `ResponseStatusException` 처리
- DefaultHandlerExceptionResolver: `IllegalArgumentException`, `HttpMethodNotSupportedException` 등 처리

하지만 Interceptor에서 발생한 예외는 위 Resolver들로는 처리되지 않기 때문에,
별도의 커스텀 `HandlerExceptionResolver`를 정의하여 예외를 잡고, 원하는 방식으로 응답을 구성할 필요가 있었습니다.
그렇지 않으면 예외는 WAS까지 전파되어 HTTP 500 오류로 응답됩니다.

---
그런데 코드 리팩터링 과정에서 Resolver가 아닌 Interceptor 내에서 바로 예외 처리를 하는 구조가 되어서, 다시 개선했습니다. 
기존: Interceptor에서 예외 상황에 맞게 리다이렉트 또는 페이지 반환
변경: Interceptor에서 예외 발생 + ExceptionResovler에서 처리

[refactor(auth): Interceptor 예외 직접 처리 방식 → Resolver 위임 방식으로 개선](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/da448f87be4af7e37821635cae2f548244b1e37a)

seovalue commented on src/main/java/roomescape/auth/infrastructure/JwtTokenProvider.java:
```diff
@@ -0,0 +1,75 @@
+    private final long validityInMilliseconds;
+
+    public JwtTokenProvider(
+            @Value("${security.jwt.token.secret-key}") String secretKey,
```
`@Value`를 통해 주입해주신 이유가 궁금해요!

yuyoungrhee replied:
사실 크게 고려하지 않고 학습테스트에 있던 내용을 참고한 것인데요, 항상 배운 것을 그대로 맹신하지 말고 어떤 방식이 더 있는지, 많은 방법 중 왜 이 방법이어야 했는지를 고려해야될 것 같네요... 코드에 제 의도가 담겨있어야 한다는 걸 까먹지 말아야겠어요🥲

SecretKey 주입 방식은 여러가지가 있습니다.
1. `@Value`: 타입 안정성 부족(런타임에서 에러 발생), 테스트코드에서 Mocking 어려움, 다른 곳에서 같은 설정이 필요할 시 중복 코드 발생 등
2. `@ConfigurationProperties`: 타입 안정성 고려, 관련 설정을 하나의 클래스로 묶어 관리 가능, 테스트 용이(설정만 바꿔 테스트)
3. 환경 변수(env) + getEnv(): 보안상 git에 노출되지 않는다는 점이 장점. Spring과 무관하게 동작

일반적으로는 `환경변수` + `@ConfigurationProperties`를 활용하여 secretKey를 안전하게 관리하고, 스프링 친화적인 방식을 유지한다고 합니다. 환경변수파일은 스프링부트와 무관하기 때문에 자동주입을 위해서는 설정이 더 필요하므로, 현재로서는 불필요하다고 생각합니다.

따라서 여러가지 방법이 있다정도만 파악하고, `@Value`를 통한 주입을 유지하기로 했습니다.

seovalue commented on src/main/java/roomescape/auth/infrastructure/JwtTokenProvider.java:
```diff
@@ -0,0 +1,75 @@
+        this.validityInMilliseconds = validityInMilliseconds;
+    }
+
+    public String createToken(String payload, Role role) {
```
JWT 토큰이 무엇이어서 사용해주신거예요?

yuyoungrhee replied:
JWT는 Json Web Token의 약자로, 사용자 인증 정보를 JSON 형태로 담아 서명한 토큰입니다.
무상태(서버는 인증 정보 저장X), Self-contained(토큰에 모든 인증 정보가 담겨있는) 등의 특징이 있습니다.

**인증 흐름**
1. 로그인 요청 (ID/PW)
2. 서버가 JWT 발급 (사용자 정보 + 서명 포함)
3. 클라이언트가 JWT 저장 (주로 쿠키나 헤더)
4. 요청 시마다 JWT 전송 (Authorization 헤더 등)
5. 서버가 JWT 검증 후 인증 처리

→ 인증된 사용자인지 서버가 기억하지 않아도, 토큰만으로 확인할 수 있습니다.

---
고려했던 다른 방식은 다음과 같습니다.
1. Basic Auth
    - 매 요청마다 Id:Password를 Base64로 인코딩하여 전송
    - 장점: HTTP 표준 방식으로, 브라우저 및 도구에서 기본 지원됨
    - 단점: 보안에 매우 취약, 비밀번호가 요청에서 노출

2. Session 기반
    - 로그인 시 서버가 세션ID 발급, 서버가 Session 저장소를 유지
    - 장점: 가장 보안성이 좋음, 요청 자체에 민감한 정보가 노출되지 않음, 서버가 인증 상태를 완전히 제어할 수 있음
    - 단점: 서버에서 세션 유지 필요 -> 확장이 어려움

3. 일반 Token 기반 (UUID 등)
    - 로그인 시 무작위 토큰 생성, DB에 유효한 토큰 저장 후 검증
    - 장점: 발급/폐기가 명확하게 관리됨, 토큰 생성,검증, 폐기 전 과정을 서버에서 통제하므로 세밀하게 제어 가능
    - 단점: Session과 같이 토큰 저장소가 필요하기때문에 Stateful함, 토큰 자체에는 의미 있는 정보가 없기 때문에 항상 서버 조회 필요

seovalue commented on src/main/java/roomescape/auth/infrastructure/JwtTokenProvider.java:
```diff
@@ -0,0 +1,75 @@
+                .setSubject(payload);
+        claims.put("role", role.name());
+        Date now = new Date();
+        Date expiry = new Date(now.getTime() + validityInMilliseconds);
```
expiry와 claim은 어떤 기준으로 정한거예요?

yuyoungrhee replied:
### expiry(만료시간) 설정 기준

expiry는 보안, 사용자 편의성, 리프레시 토큰 도입 여부에 따라 설정됩니다.

보안 측면
- 짧게 설정하면: 탈취 시 피해를 줄일 수 있음
- 길게 설정하면: 같은 토큰이 오래 지속되어 보안에 취약해질 수 있음

편의성 측면
- 만료 시 로그인을 다시 해야 하므로, 너무 짧으면 사용성이 떨어짐 (예: 5분마다 로그인)
- 서비스 특성에 따라 적절한 만료 시간을 선택해야 함

리프레시 토큰 도입 여부
- 리프레시 토큰이 있다면: 액세스 토큰은 짧게 설정해 보안 확보
  - 리프레시 토큰은 길게 설정해 사용자 편의 보장
- 이 구조를 통해 보안과 편의성을 모두 확보할 수 있음

현재는 엑세스토큰만 있는 구조이므로, 1시간으로 설정했습니다.

### claim (토큰에 담는 정보) 기준

현재 Claim 정보
- sub: 토큰의 제목, 보통 식별할 수 있는 정보가 들어감 (ex: userId, email 등)
- role: 권한 판단 (ex: ADMIN, USER 등)
- iat, exp: 발급/만료시간

**Claim에 어떤 정보를 넣어야 하는가? -> 토큰 하나로 인증과 인가를 판단할 수 있어야한다"는 원칙에 따라 결정**

각 Claim의 목적
- sub: 사용자 식별
- role: 권한 판단(인가에 필요)
- iat: 발급 시점 기록 (토큰이 너무 오래된 건 아닌지 확인)
- exp: 만료 시점 설정: 토큰 유효시간을 설정(보안을 위해 반드시 필요)

seovalue commented on src/main/java/roomescape/auth/presentation/AdminAuthorizationInterceptor.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
```
Interceptor와 HandlerArgumentResolver는 어떤 차이예요?

yuyoungrhee replied:
### Interceptor와 HandlerArgumentResolver 차이

구분 | Interceptor | HandlerMethodArgumentResolver
-- | -- | --
실행 시점 | 컨트롤러 호출 전/후 전체 요청 처리 | 컨트롤러 메서드 파라미터 바인딩 시점
목적 | "이 요청을 계속 진행시킬 것인가?" | "컨트롤러에 어떤 데이터를 넣어줄까?"
목적 예시 | 인증, 인가, 로깅, 요청 차단 등 | 파라미터 주입 (`@LoginMember` 등)
적용 범위 | 전체 요청 (URL 기준 매핑) | 특정 파라미터 타입 또는 어노테이션 기준
결과 반환 | true/false로 흐름 제어 | 파라미터 객체 반환

Interceptor: 인증 + 인가 (요청 자체 허용 여부 판단)
- 토큰 존재 여부 및 유효성 검증
- 토큰에서 사용자 정보 추출 (ex. userId, role 등)
- 사용자가 로그인되어 있는지 확인
- 요청 URI/메서드에 대한 권한(Role) 검사
- 통과 시 사용자 정보를 request.setAttribute()에 저장

Resolver: 이미 인증된 정보 주입 (컨트롤러 파라미터 주입)
- request에 저장된 사용자 정보를 꺼냄
- 또는 토큰에서 직접 사용자 ID 파싱 (Interceptor가 안 했을 경우 대비)
- 컨트롤러 파라미터로 안전하게 전달 (@LoginMember Member member)
- 방어적 차원에서 null 체크 또는 토큰 유효성 재확인

yuyoungrhee replied:
Interceptor와 Resolver에 대해 정리하면서 제가 잘못 책임을 분리한 부분이 있는 것 같아 수정했습니다.

[refactor(auth): 인증 책임을 Interceptor로, 주입 책임을 Resolver 로 분리](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/a8ded7ba98e1b9196adf367b73c2800db5bd0c9b)

**기존 코드**
- Interceptor에서 인증/인가 검증 후 요청 진행 여부만 반환
- Resolver에서 토큰 정보 재인증 및 주입

**변경**
- Interceptor에서 인증/인가 검증 후 사용자 정보를 request.setAttribute()로 저장
- Resolver는 request의 정보를 활용하여 중복 인증 로직 없이 주입
- 인증 책임은 Interceptor에, 주입 책임은 Resolver로 책임을 명확히 분리

seovalue replied:
좋네요! resolver는 주입역할만 하니깐 오히려 명확해진 것 같기도 하네요!
다만 궁금한 점이 있는데 해당 코드에 남겨볼게요~!

seovalue commented on src/main/java/roomescape/auth/presentation/AuthenticatedMember.java:
```diff
@@ -0,0 +1,11 @@
+
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface AuthenticatedMember {
```
해당 Interface를 사용해주시게 된 이유가 궁금해요~

yuyoungrhee replied:
컨트롤러 파라미터에 로그인 사용자 라는 특정 객체를 주입하고 싶었기 때문에, `HandlerMethodArgumentResolver` + 커스텀 어노테이션을 사용했습니다.

`@interface`는 자바 문법상 어노테이션 정의 키워드입니다. 어노테이션은 자바에서 특별한 인터페이스 형태로 취급되며, 컴파일 시에는 `java.lang.annotation.Annotation`을 구현하는 클래스로 변환됩니다.

`@RequestParam`이나 `@RequestBody` 등은 스프링이 제공하는 여러 기본 ArgumentResolver들이 처리해줍니다. 지금처럼 커스텀 어노테이션을 사용하는 경우 따로 `ArgumentResovler`를 정의해야 합니다.

seovalue commented on src/main/java/roomescape/auth/presentation/AuthenticatedMemberArgumentResolver.java:
```diff
@@ -0,0 +1,51 @@
+        String token = extractor.extract(request)
+                .orElseThrow(() -> new AuthorizationException(LOGIN_REQUIRED));
+
+        Member member = authService.findMemberByToken(token);
+        return new LoginMember(member.getId(), member.getEmail(), member.getName());
```
memberID 만을 반환할 수도 있을 것 같은데 email과 name 모두 반환하는 이유가 궁금해요!

yuyoungrhee replied:
이 부분은 착오가 있었습니다... 미션 명세에서 email과 name을 모두 주는 줄 알았어요! 다시 확인해보니 name만 주도록 되어있네요. 

/login/check : 사용자 이름을 상단에 표시해주기 위해 호출 됨, name 필요
/reservations : 사용자 id를 통해 예약을 생성함, id 필요

-> LoginMember가 id, name을 가지도록 변경 (email 제외)
[refactor(LoginMember): 불필요한 정보인 email 제거, api 명세 업데이트](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/0a66d1e2a4e11ab8117333ed27a64c5746530136)

seovalue commented on src/main/java/roomescape/auth/presentation/TokenLoginController.java:
```diff
@@ -0,0 +1,51 @@
+
+        TokenResponse tokenResponse = authService.createToken(tokenRequest);
+
+        Cookie cookie = new Cookie("token", tokenResponse.accessToken());
+        cookie.setHttpOnly(true);
+        cookie.setPath("/");
+        response.addCookie(cookie);
```
cookie에 세팅하는 로직을 한곳에서 관리해도 좋겠네용

seovalue replied:
그나저나 Cookie가 뭐예요?

yuyoungrhee replied:
좋은 제안 감사합니다! Cookie 세팅 로직을 한 곳에서 관리하도록 했어요. 책임을 분리하고 재사용 및 중복 코드 제거에 도움이 될 것 같습니다.

[refactor(login): CookieProvider를 도입하여 토큰 쿠키 설정 로직을 컨트롤러에서 분리](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/6d35e6062f7d5a0d0e7da316c6f6e13aa0fa8c98)

---

Cookie란
- HTTP 요청과 응답 사이에 상태 정보를 유지하기 위해 서버가 클라이언트에 저장시키는 데이터
- 클라이언트 브라우저에 저장됨
- 클라이언트는 이후 동일 도메인 요청마다 해당 쿠키를 자동으로 함께 전송
- 사용 예시: 로그인 상태 유지, 사용자 식별, 장바구니 상태 유지 등

현재 사용하고 있는 Cookie 용도
- JWT 토큰을 Cookie에 저장함으로써 인증 상태 유지

seovalue commented on src/main/java/roomescape/member/application/MemberService.java:
```diff
@@ -0,0 +1,33 @@
+
+    public MemberDto getMemberById(Long id) {
+        Member member = memberRepository.findById(id)
+                .orElseThrow(() -> new NotFoundException("memberId", id));
```
notfound의 메시지로 id 넘기는 것 👍

seovalue reviewed (changes requested):
안녕하세요 멍구~! 미션 빠르게 진행해주셨네요ㅎㅎ
코멘트 남기다보면 길어질 것 같아서 우선 한번 이야기해보고, 
다음 제출에 이어서 테스트 및 구현단에 대해 더 이야기나눠보면 좋을 것 같아요!
확인해보시고 다시 리뷰 요청 주세요!👍

YuyoungRhee committed:
refactor: DomainException의 위치를 shared로 이동

- 의존성 개선을 위해 domain 관련 예외가 외부를 알지 못하도록 변경

YuyoungRhee committed:
refactor(auth): Interceptor 예외 직접 처리 방식 → Resolver 위임 방식으로 개선

Interceptor에서 예외를 발생시키고 HandlerExceptionResolver에서 처리하도록 구조 변경

YuyoungRhee committed:
refactor(LoginMember): 불필요한 정보인 email 제거, api 명세 업데이트

YuyoungRhee committed:
refactor(auth): 인증 책임을 Interceptor로, 주입 책임을 Resolver 로 분리

- 관리자 권한을 검사하는 admin Interceptor 외 BasicAuthorizationInterceptor도 구현
- Interceptor에서 인증된 사용자 정보를 request에 저장하도록 변경
- Resolver는 request의 사용자 정보를 활용하여 중복 인증 로직 없이 주입

YuyoungRhee committed:
refactor(login): CookieProvider를 도입하여 토큰 쿠키 설정 로직을 컨트롤러에서 분리

YuyoungRhee commented:
조앤 리뷰 감사합니다! 코멘트 모두 답변했습니다.

변경 사항
- DomainException 위치 고민 -> shared/domain/exception으로 domain 공통 패키지 생성하여 위치시킴
- Interceptor와 ExceptionReolver의 책임 분리 -> Interceptor에서 예외 throw, ExceptionResolver에서 처리 및 리다이렉트
- ArgumentResolver에서 중복으로 토큰 인증하는 구조 개선 -> Interceptor에서 인증/인가 + Argument Resolver에서는 인증된 사용자 주입
- Cookie 세팅 로직 클래스 분리 -> CookieProvider 구현, TokenController에서는 CookieProvider를 주입받아 사용

seovalue commented on src/main/java/roomescape/config/WebMvcConfiguration.java:
```diff
@@ -40,7 +45,10 @@ public void addArgumentResolvers(List<HandlerMethodArgumentResolver> resolvers)
-        registry.addInterceptor(adminInterceptor)
+        registry.addInterceptor(basicAuthInterceptor)
+                .addPathPatterns("/**")
+                .excludePathPatterns("/", "/login", "/css/**", "/js/**", "/image/**","/themes/ranking", "/favicon.ico");
```
조회와 같이 token이 필요없는 경우 매번 하나하나 추가해줘야할 것 같은데, 이 부분에 대해서 어떻게 생각해요?

yuyoungrhee replied:
보안 측면에서는 우선 모든 요청을 막아두고 제외되는 경우만 명시해주는 게 좋다고 생각해요. 실수로 빠뜨려서 **관리자 페이지 공개하기 vs 공개api 접근 못하기** 라고 한다면 후자를 선택할 것 같거든요..

그런데 확실히 이런 식의 구현은 복잡하고 가독성이 좋지 않은 것은 맞는 거 같아, 몇가지 방법을 생각해보았습니다..

### 1. 공개된 api임을 나타낼 수 있는  `/public` prefix 붙이기
`InterceptorRegistry`에는 `excludPatPatterns("/public/**")` 이렇게 추가하는 식으로 관리해볼 수 있을 것 같습니다.
그런데 이러면 Controller 자체가 하나의 도메인에 대해서 둘로 나뉘어야 합니다. 

예를 들자면 `PublicReservationController` / `MemberReservationController`
이런 식으로요.. 이렇게 된다면 아직 구조가 크지도 않은데 같은 도메인에 대한 컨트롤러가 나뉘어진다는 것에서 복잡도가 증가하는 단점이 있습니다.

### 2. token이 필요없는 경우를 나누어 잘 관리하기
token이 필요없는 경우는 크게 다음과 같습니다.
- 정적 리소스 호출(css, js 등)
- 로그인, 회원가입 등의 페이지 및 api
- 로그인하지 않아도 사용할 수 있는 공개된 페이지 및 api

이를 나누어서 List로 관리한다면 가독성 있게 관리할 수 있을 것 같습니다. 현재는 복잡도를 늘리고싶지 않아 2번으로 해결해보았습니다.

[refactor(config): 인증 제외 경로를 의미별로 리스트 분리하여 관리하도록 개선](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/504da8040924b93b96c2279f8e5ab8ee354e7519)

```java
    @Override
    public void addInterceptors(InterceptorRegistry registry) {

        registry.addInterceptor(basicAuthInterceptor)
                .addPathPatterns("/**")
                .excludePathPatterns(STATIC_API_LIST)
                .excludePathPatterns(PUBLIC_API_LIST)
                .excludePathPatterns(PUBLIC_AUTH_API_LIST);
        registry.addInterceptor(adminAuthInterceptor)
                .addPathPatterns("/admin/**");
    }
```

제가 고려하지 못한 다른 좋은 방법이 있을까요? 검색으로 관련 내용을 찾아보기는 쉽지 않았습니다.

seovalue replied:
오홍! 한데 모아서 관리해주는 형태로 적용해보셨군요👍🏻 저는 이 방법도 좋다고 생각해요.

추가적으로 제가 당장 떠오르는 방법은, 
우선 어드민 인터셉터는 지금을 유지하는게 좋을 것 같고요.

그 외 유저 기능들에 대해서는 기본적으로 토큰을 받을 수 있게 하되(=인터셉터를 다 타는거죠) 인터셉터 레벨에서는 토큰으로 파싱된 멤버가 있다면 attribute에 주입해주고, 그렇지 않으면 null을 주입해준 뒤, 사용하는 쪽에서 전달받은 member가 null일 때 적절한 처리를 해주는 방법도 있을 것 같아요.

이렇게 되면 401에 대해서 컨트롤러 레벨에서 처리하게 되어서 응집도가 떨어질 순 있을 것 같네요ㅎㅎ 

멍구가 고민해보시고 더 좋은 아이디어 떠오른다면 저도 알려주세요!👍🏻

seovalue commented on src/main/java/roomescape/reservation/presentation/controller/ReservationController.java:
```diff
@@ -0,0 +1,57 @@
+    @PostMapping
+    public ResponseEntity<ReservationDetailResponse> createReservation(
+            @Valid @RequestBody ReservationRequest request,
+            @AuthenticatedMember LoginMember member) {
```
```suggestion
            @AuthenticatedMember LoginMember member) {
```
위 부분에 대해서, 이번에 request.setAttribute 형태로 변경된 김에 Custom Annotation 없이 `@RequestAttribute("loginMember") LoginMember member` 이런 형태로도 사용해볼 수 있을 것 같거든요! 이 부분은 어떻게 생각하시는지 궁금해요!!

yuyoungrhee replied:
오  `@RequestAttribute`는 처음 알았어요. HTTP 요청(request)의 attribute에 저장된 값을 컨트롤러 메서드 파라미터로 바인딩할 때 사용하는 어노테이션이군요!

CustomAnnotation과 비교해봤을 때 코드량을 제외한다면 약간의 단점이 있는 것 같습니다.

**인증에 관한 정보를 Controller가 알아야합니다.** 
request에 "loginMember"로 저장되어있다는 것을 알아야 하기 때문입니다. 상수화로 관리해볼 수 있겠지만 Controller가 인증 관련 정보를 알아야한다는 사실 자체가 단점인 것 같습니다.

그리고 개인적으로 인증정보가 "꺼내 쓰는" 느낌보다, "주입 되는" 느낌을 주는 것이 관심사가 더 잘 분리되었다고 느낍니다. 객관적이지 못한 것 같지만, 그만큼 `@RequestAttribute`도 사실상 잘 동작하더군요. 이 둘 중에 선택은 개인적인 선호일까요? 제가 고려하지 못한 장단점이 또 있다면 알려주세요!

seovalue replied:
커스텀 어노테이션을 사용한다면 연결된 리졸버에서 추가적인 로직을 수행할 수 있는 장점도 있을 것 같아요! 그게 필요없다면 지금은 requestAttribute를 써도 상관은 없을 것 같고용ㅎㅎ

seovalue commented on README.md:
```diff
@@ -1,6 +1,34 @@
 # 방탈출 사용자 예약 애플리케이션
 
-사용자가 예약 가능한 시간을 조회하고, 예약할 수 있습니다.
+## 주요 기능
```
전체적으로 패키지 구조 구성에 대해서 이야기해주실 수 있을까요?
도메인 하위로 구분된 거라고 생각했는데, application, presentation은 또 아닌 것 같아서 정확한 기준이 뭔지 궁금해서요

yuyoungrhee replied:
도메인 뜻이 여러가지라 설명이 명확하지 못했던 것 같습니다.

1. 계층으로서의 도메인: 도메인 모델과 비즈니스 규칙을 담는 계층
2. 문제 영역으로서의 도메인: 소프트웨어가 해결하고자 하는 문제 영역 또는 구현해야하는 대상

여기서 저는 도메인 단위 별로 패키지를 구성했다고 했는데, 여기서 도메인은 2번의 의미였습니다. 즉, 해결하고자하는 기능 별로 패키지를 구성했습니다.

```
도메인(Reservation, Theme, ReservationTime 등 - 2번의 뜻)
ㄴpresentation
ㄴapplication
ㄴdomain
ㄴinfrastructure
```

이런 식으로 구성했습니다.

이렇게 구성한 이유는 변경의 단위를 패키지별로 가져가기 위해서입니다. 현재 각 도메인(기능)이 각자의 Controller, Service, Domain, Infra를 가지고 있습니다.
예를 들어 time에 기능이 추가된다면, 모든 패키지가 아니라 time 패키지만 변경되길 바라는 마음에서 패키지 구성을 이렇게 했습니다.

---
추가적으로 config, exception은 global 패키지로 묶었습니다. 
[refactor: 패키지 구조 정리 - global 및 auth](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/e8707d6ab988e5b37d9be5013dfc509c0cfb2f82)

shared.domain.exception을 구성한 것과 같이 모든 도메인 및 계층에서 공유된다는 의미에서 global로 네이밍을 두었습니다.

- global.exception: 표현, 응용 계층 등 공통적으로 사용하는 예외 처리 및 형식
- global.config: Spring 설정 등 어느 도메인에도 종속되지 않는 전역 설정

seovalue commented on src/test/java/roomescape/AbstractRestDocsTest.java:
```diff
@@ -0,0 +1,52 @@
+
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
+@ExtendWith(RestDocumentationExtension.class)
+public abstract class AbstractRestDocsTest {
```
RestDocs 세팅하는 데 어려운 점은 없으셨나요?

yuyoungrhee replied:
#### 1. 개발 시 build/docs/에 있는 정보를 복사해와야하는 문제
우선 정적 리소스는
- 개발 중에는 `src/main/resources/static` 안의 정보가 반영된다.
- 배포 중에는 `build/resource/main/static` 안의 정보가 반영된다.

    그리고 `restdocs-asciidoctor` 라이브러리는 기본적으로 테스트 실행 후 스니펫이 **`build/docs/asciidoc/` 디렉토리에 저장됩니다.** 

    그러나 개발 중에는 `src/main/resources/static`에 이 정보가 들어있어야 하기 때문에, 직접 복사를 하거나 따로 설정이 필요합니다. 뭔가 테스트하면 자동으로 딱 복사가 되게 하고 싶었는데, 이렇게는 따로 설정이 불가능하다고 하네요.. 

     차선책으로 build 시 `build/docs/asciidoc` 안의 정보가 `src/main/resoruces/static/docs`로 복사되게 하여 반자동화 하였습니다.

**2. RestAssured filter에 필요한 공통 설정**
어떤 정보를 filter 걸어야하는지 처음에는 알기 어려워서, 한 번씩 돌려보고 고치고 하는 과정을 거쳤습니다. 어려운 점 이라고하기에는 애매하네요..ㅎㅎ Response에 자동으로 API 명세에는 필요 없는 정보가 포함되어서 remove 시키거나, JSON 형식이 한 줄로 나오는 것을 `prettyPrint()` 등으로 해결했습니다.

seovalue commented on src/test/java/roomescape/auth/application/AuthServiceTest.java:
```diff
@@ -0,0 +1,107 @@
+
+        given(memberRepository.findByEmail("email@test.com")).willReturn(Optional.of(member));
+        String payload = String.valueOf(memberId);
+        given(jwtTokenProvider.createToken(payload, member.getRole())).willReturn("token-value");
+
+        TokenResponse response = authService.createToken(request);
+
+        assertThat(response.accessToken()).isEqualTo("token-value");
```
given에서 세팅된 값이 결국 반환되는 것 같은데, 지난번에 이야기했던 의미있는 테스트와도 연관된 것 같아요! 이 부분에 대해서 멍구의 의견은 어때요?

yuyoungrhee replied:
앗, AuthServiceTest에서는 신경쓰지 못했네요. verify를 추가했습니다!
[test(application): 서비스 테스트 보강](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/153360c6399b21fa1ba06024816dfced2da61dc7)

assertThat도 남겨두었는데, 그 이유는 최종 반환값이 기대한 형식인지를 확인하고, 내부를 검증했을 때 그대로 나오는지도 의미있다고 생각했기 때문이에요. verify로 행위만 검증한다면 authService.createToken의 반환값은 테스트할 필요가 없게 되는데, 반환값이 Member로나오는지, MemberDto로나오는지, 혹은 다른 이상한 형식으로 나오는지도 테스트할 필요가 있다고 느꼈기 때문입니다.

seovalue commented on src/test/java/roomescape/auth/infrastructure/JwtTokenProviderTest.java:
```diff
@@ -0,0 +1,50 @@
+import org.junit.jupiter.api.Test;
+import roomescape.member.domain.Role;
+
+class JwtTokenProviderTest {
```
👍

seovalue commented on src/test/java/roomescape/auth/presentation/TokenLoginControllerTest.java:
```diff
@@ -0,0 +1,72 @@
+
+    @BeforeEach
+    void cleanDatabase() {
+        jdbcTemplate.execute("SET REFERENTIAL_INTEGRITY FALSE");
+        jdbcTemplate.execute("TRUNCATE TABLE members");
+        jdbcTemplate.execute("ALTER TABLE members ALTER COLUMN id RESTART WITH 1");
+        jdbcTemplate.execute("SET REFERENTIAL_INTEGRITY TRUE");
```
cleanDatabase를 매 테스트 실행 전에 실행될 수 있도록 한군데서 관리해봐도 편할 것 같네요. Database Cleaner와 같은 형태로.?

yuyoungrhee replied:
정말 좋은 생각 같습니다!! 매 테스트 실행 전에 필요한 건 아니라서, DatabaseCleaner를 빈으로 주입받아 원하는 테스트에 databaseCleaner.clean() 시켜줬습니다.

[refactor(test): DatabaseCleaner 유틸 클래스로 데이터 초기화 로직 분리](https://github.com/woowacourse/spring-roomescape-member/pull/253/commits/b70f95fcc25b21f2997bd738ca17cd4007aadd48)
```java
@BeforeEach
void clean() {
    databaseCleaner.clean();
}
```

seovalue replied:
단번에 적용하시다니 멋지네요

seovalue commented on src/test/java/roomescape/auth/presentation/TokenLoginControllerTest.java:
```diff
@@ -0,0 +1,72 @@
+                .when().post("/login")
+                .then().log().all()
+                .statusCode(200)
+                .header("Set-Cookie", Matchers.containsString("token="));
```
👍

seovalue commented on src/test/java/roomescape/reservation/application/ReservationQueryServiceTest.java:
```diff
@@ -0,0 +1,72 @@
+
+    @DisplayName("검색 조건이 있으면 해당 조건으로 예약 상세 정보를 조회한다.")
+    @Test
+    void getReservationDetails_withCondition() {
+        // given
+        ReservationSearchCondition condition = new ReservationSearchCondition(1L, null, null, null);
+        given(reservationQueryDao.findByCondition(condition))
+                .willReturn(List.of(mock(ReservationDetailData.class)));
+
+        // when
+        reservationQueryService.getReservationDetails(condition);
+
+        // then
+        verify(reservationQueryDao).findByCondition(condition);
+    }
```
이번에 Verify를 잘 활용해주셨네요!
또 궁금한 점은 해당 검색 조건에 해당하는 예약이 나오는지에 대한 테스트를 하려면 어떻게 하는게 좀 더 적합할까요?

yuyoungrhee replied:
어렵네요.. 해당 검색 조건에 해당하는 예약이 나오는지에 대한 테스트를 하려면 Repository까지 연관되는 Fake 혹은 Service-DB 통합 테스트를 하는게 더 좋을 것 같아요.
반환값 검증을 Stub로 하더라도 정해진 결과를 줄 수밖에 없기에, 검색 조건에 해당하는 예약이 나오는지는 검증하기 힘들 것 같아요.

현재 QueryDao에 대한 테스트가 작성되어있는 상태여서, Service에서는 입력 형식, QueryDao 메서드 실행 여부 verify, 반환값 형식 등만 확인해도 충분히 신뢰할 수 있다고 생각합니다. 굳이 Service에서 QueryDao까지의 통합테스트를 한 번 더 진행하는 건 너무 과한 것 같아요. 어떻게 생각하시나요?

seovalue replied:
저도 지금 구조에서는 주신 의견에 동의해요! 사실상 필터링 로직이 db 쿼리와 강결합이어서ㅎㅎ 만약 서비스레이어에서 필터링이 비즈니스로직으로 있었다면 그때는 말씀주신 스텁 등을 활용해보면 좋을 것 같아요!👍🏻

seovalue reviewed (changes requested):
안녕하세요 멍구! 빠르게 수정해주셨네요!
제 질문에 깊이있게 답변 남겨주셔서 저도 배울 수 있었어요 감사해요 👍 
기존 코멘트에 대한 답변 및 몇가지 코멘트 남겼는데 확인 후 리뷰 요청 주세요~!
다음 요청 땐 머지할 것 같아요ㅎㅎ

YuyoungRhee committed:
docs: API 명세 업데이트

YuyoungRhee commented:
@coderabbitai review

YuyoungRhee commented:
@coderabbitai full review

YuyoungRhee committed:
refactor(config): 인증 제외 경로를 의미별로 리스트 분리하여 관리하도록 개선

- 정적 리소스, 공개 API, 인증 관련 API로 나누어 관리

YuyoungRhee committed:
refactor: 패키지 구조 정리 - global 및 auth

- global 성격을 지닌 패키지를 global패키지에 모음
- auth의 dto 패키지를 auth 루트에 두지 않고 각 계층에 사용되는 곳으로 이동

YuyoungRhee committed:
test(application): 서비스 테스트 보강

- verify, assertTha를 활용하여 행위, 상태 모두 검증하도록 보강함

YuyoungRhee committed:
refactor(test): DatabaseCleaner 유틸 클래스로 데이터 초기화 로직 분리

YuyoungRhee commented:
조앤 리뷰 감사합니다! 전체적으로 다시 생각해보면서 개선할 수 있었어요.
마지막 요청이 될 것 같은데, 의문점이나 더 개선할 수 있는 부분들은 코멘트 남겨주시면 충분히 고민해서 다음 미션에 적용할 수 있도록 해볼게요🤩

seovalue commented on README.md:
```diff
@@ -32,3 +60,7 @@
     - [X] 최근 일주일을 기준으로 하여 해당 기간 내에 방문하는 예약이 많은 테마 10개를 확인합니다.
         - [X] `GET - /` 요청 시 인기 테마 페이지를 응답합니다.
+- [x] 로그인 기능 구현
+  - Cooki, `Argument Resolver` 활용하여 구현
```
마지막 질문!! 멍구가 생각하시는 인증과 인가에 대해서 설명해주실 수 있으실까요?
(gpt금지하구 진짜 이해한 내용을 작성해주시면 멍구에게 더 도움이 될듯해요~!)

yuyoungrhee replied:
제가 이해한 바로는
인증: 너 누구야!
인가: 누구인지는 알았는데, 그걸 할 수 있는 사람이야?

---
비유적으로 표현하면 와닿는 거 같아서 이런식으로 설명하고 싶네요. 

예를 들어 어떤 식당에 "사람만 입장 가능"하다는 규칙이 있다면, 입장할 때 사람 인증서(있다고 가정합니다..ㅎㅎ)를 제출해서 내가 사람임을 증명해야 합니다.  

그리고 식당에서 술을 주문하는 행위를 하려면 "만 19세 이상"이어야 합니다.  
이때는 민증을 제시해서 생년월일을 확인합니다.

술을 시킬 때 다시 "사람 인증서"를 보여줄 필요는 없습니다. 이미 입장할 때 인증되었기 때문이에요.

여기서 "사람"인지를 확인하는 과정을 인증, 술을 시키는 행위가 가능한 사람인지를 확인하는 과정을 인가라고 할 수 있겠네요!

seovalue reviewed (approved):
멍구! 남겨주신 질문에 대해 제 생각들 남겨봤어요. 어려운 미션 빠르게 진행해주시느라 고생 많으셨어요~! 질의응답 하는 것도 재밌었어요! 
앞으로 남은 미션들도 화이팅이에요!
